### PR TITLE
Sickbeard/Sickgear/Sickrage: Init and module

### PR DESCRIPTION
**Pull request by rembo10**
###### Motivation for this change
This is an attempt to unify #46538, #46214 and #46340. Added packages for sickbeard, sickgear and sickrage, and added a nixos module to specify which package to use.
@sterfield: I hope it's ok that I didn't merge your commit - since I made some changes to your package based on comments from @worldofpeace on #46514. I kept you as the maintainer for Sickrage though

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.bool; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.path; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.str; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Should be "${dataDir}/config.ini and be sure to add a rec above so you can reference dataDir properly, I believe.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.ints.u16; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.str; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Would love to see a type = types.str; here as well.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: In the options you allow the user to specify username but don't act on it. If I wanted to run sickgear as my own user I could not. See the user and group creation from here as an example:

https://github.com/NixOS/nixpkgs/blob/980cbff93c20b5dcd4ac63384db84490f5e07a45/nixos/modules/services/misc/redmine.nix

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Should be group instead of "sickgear".

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Should be dataDir.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: Same comment as user. See group in linked example.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [inline nixos/modules/services/networking/sickbeard.nix]: According to #45638:

"I don't remember where we have that documented, but service descriptions should be upper-case-first for every word."

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/sickbeard.nix
Comment (rembo10) [reply nixos/modules/services/networking/sickbeard.nix]: types.str or types.path?

Comment (aanderse) [reply nixos/modules/services/networking/sickbeard.nix]: Based on a quick search I believe types.path can only refer to an actual directory and not a file, but I'm not certain of this.

I believe the answer is types.str but would appreciate being corrected by anyone of this is not the case.

Comment (rembo10) [reply nixos/modules/services/networking/sickbeard.nix]: I think (according to [this](https://github.com/NixOS/nixpkgs/blob/80d7a4c87281b4fb41105e4bd22d4d600fdbf153/lib/types.nix#L236)) it means it's an absolute path? Which I think would be better here as, at least with sickgear, it seems like it puts relative paths relative to /. So it was failing when I tried it with types.str with a relative path. At least types.path enforces an absolute path

**Comment by rembo10** [comment]
@aanderse : thanks for all the comments. I don't know how I missed the user/group stuff. Anyways can you check the new commits and see if they're alright? I'll make similar changes to the headphones (#46514) module.

File: nixos/modules/services/networking/sickbeard.nix
Comment (aanderse) [reply nixos/modules/services/networking/sickbeard.nix]: Good to know, thanks!

**Review by aanderse — APPROVED** [review]
Looking good to me.

**Comment by sterfield** [comment]
Hey all, 

@rembo10 No worries, I agree with you, it's better to have all similar packages behind the same service. I suppose that, as they are all forks of sickbeard, they work the same.

However, I don't understand why we don't leverage Systemd's [`dynamicUser`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser=) and all relative options (`RuntimeDirectory`, `StateDirectory`, etc…).

I did that in my PR (see https://github.com/NixOS/nixpkgs/pull/46340/files#diff-f653b9eaf598496872612beefd0cb752R32).

The beginning of `ids.nix` state that:

> # We only add static uids and gids for services where it is not feasible
> # to change uids/gids on service start, in example a service with a lot of
> # files.

And I don't think this is the case here.

What do you think ?

**Comment by aanderse** [comment]
> However, I don't understand why we don't leverage Systemd's [`dynamicUser`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser=) and all relative options (`RuntimeDirectory`, `StateDirectory`, etc…).

@sterfield Interesting. Does sickbeard not store download files under the directory specified? How would that work with DynamicUser?

I have transmission running on my system and since there is a dedicated user I can add the transmission group to my own user accounts extra groups so I have access to the downloaded files.

**Comment by sterfield** [comment]
Well, I looked at systemd's documentation page and found lots of interesting stuff.

Here, `dynamicUser` will not help, as Sickbeard/Sickrage/whatever searches for a corresponding line in `/etc/passwd`, but we can only declare a simple `user` and let the system decides the uid/gid. There's no point (IMHO) to force a specific uid/gid. I did the code in my PR (https://github.com/NixOS/nixpkgs/pull/46340/files#diff-f653b9eaf598496872612beefd0cb752R39) (actually, I declared only `sickrage` as a user and `nogroup` as a group. We could create an option to create a group if there's a specific need.)

As for the RuntimeDirectory, it's by default under `/var/lib/sickrage` which is the correct place. The only drawback I can see is the user cannot specify a data directory, but I don't think it's a problem, as it's only the place where the config file and the db is stored.

**Comment by sterfield** [comment]
> @sterfield Interesting. Does sickbeard not store download files under the directory specified? How would that work with DynamicUser?

As far as I can remember, no. Sickbeard is storing its db in the `data` path, but download folders are then added after by the user.

**Comment by rembo10** [comment]
I think it's useful to give them dedicated ids since they're moving files around. For example, I add sickbeard to the 'storage' group, which has write access to a mounted drive

**Comment by sterfield** [comment]
> I think it's useful to give them dedicated ids since they're moving files around. For example, I add sickbeard to the 'storage' group, which has write access to a mounted drive

But an non-dedicated id for sickrage / sickbeard will allow you to do that, no ?

**Comment by rembo10** [comment]
Ah would it? Ok, sorry I think maybe I misunderstood. I'm happy to make the change as long as it'll work with my setup 😂😅

**Comment by sterfield** [comment]
It has to be tested, but from my point of view:
- a `sickbeard` user is created along with the service
- you declare `sickbeard` as member of your `storage` group
- you declare your mountpoint with the `storage` group

and that should be it.

One additional option we could create is a list of group user `sickbeard` belongs to. By default, we keep `nogroup` but if the user add some groups, then we add `sickbeard` to those groups.

**Comment by rembo10** [comment]
I'm still doing a bit of research but I stumbled across [this comment](https://github.com/PurpleI2P/i2pd/issues/1091#issuecomment-362841840) that seems to suggest that it won't work for my use case (and I imagine many others') as these services are usually used in conjunction with transmission, sabnznd, etc. Sorry - I'm still trying to wrap my head around all of this so I might be missing something totally obvious. Reading the systemd link above, it seems like you can pass a list of groups to run the service as, but do they also have access to directories outside of /var/lib/service? Or would those directories need to be mounted?

**Comment by sterfield** [comment]
That's what I mentioned in one of my comment:
- `dynamicUser` can be used, but you don't have a "real" user, only created by systemd at service boot, and deleted at service stop. I _can_ be used here, but Sickbeard is checking user's password in `/etc/passwd`, which doesn't exist in that case and the application crashes (I tested it with Sickrage, so I assume that it's the same for other forks)
- `runtimeDirectory` can be used here, but as stated in the comment you linked, the directory is created under `/var/lib/<some directory>` and you don't have the choice for the location.
- creation of a specific user using Nixos `user` option creates a regular user.

So we can do three things here:
- creation of a `user` without fixed uid/gid and use of `runtimeDirectory`. We don't know the uid/gid chosen by Nixos and the datadir will be handled by systemd under `/var/lib/sickbeard` (for example).
  - Advantage: don't have to manage the datadir yourself, and let nixos decides the uid/gid. `ids.nix` will be less cluttered.
  - Disadvantage: you don't have a fixed uid/gid, but I don't see any impact here. The datadir is also fixed, preventing the user to chose his own datadir. That's fine by me, as I don't change its content manually, but it _could_ be an issue for some user that wants a specific folder or to use a specific `config.ini`
- creation of a `user` without fixed uid/gid and creation of the datadir by Nixos install phase:
  - Advantage: datadir can be customized
  - Disadvantage: more things to manage at Nixos level
- creation of a `user` with fixed uid/gid and creation of the datadir by Nixos install phase:
  - Advantage: fixed uid and datadir can be customized
  - Disadvantage: more things to manage at Nixos level and `ids.nix` will have more ids declared

So ultimately, there's no big differences.
 - if we want to customize `datadir`, then let's do it, but if there's no added value, we can let systemd handle this.
- we can declare a specific uid/gid for `sickbeard` user, but I don't see any added value here.

**Comment by rembo10** [comment]
I'm not sure I know enough about this to make the decision as I'm quite new to nix / nixos so I'll defer to your guys' judgement on this

**Comment by sterfield** [comment]
I still don't understand why you want to fix uid/gid in `ids.nix` instead of using `user` option and let the system decide the uid/gid itself.

But the codes looks good to me, so if you want to merge it like that, then go for it !

File: nixos/modules/services/networking/sickbeard.nix
Comment (sterfield) [inline nixos/modules/services/networking/sickbeard.nix]: I don't think the nixos module belongs here. Sonarr, couchpotato modules are stored under `nixos/modules/services/misc` so I think sickbeard should be put there as well.

**Review by sterfield — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/sickbeard.nix
Comment (rembo10) [reply nixos/modules/services/networking/sickbeard.nix]: I think your right. Fixed.

**Comment by rembo10** [comment]
I'll leave it up to you guys to decide about the id situation. Just let me know and I can make changes

**Comment by aanderse** [comment]
@sterfield Just reread your options... Is the only difference between option 2 and 3 the predefined ids?

If so I think I'm indifferent between the two. I still think there would be value in allowing the user to specify the user and group the service runs as, but I don't know enough about sickbeard to say.

It sounds like @rembo10 and I are leaving the decision to you.

**Comment by sterfield** [comment]
So, I'm all for letting the user chose user / group to handle the service. The code you did is really cool. The only concern I have is the use of predefined ids, where I don't see really the need here. In addition, some comments in the file mentioned to not go above uid 399, and we are getting close to that.

So here's my thoughts:
- remove the predefined ids from the current code and we push the code like that.
- in the meantime, I'll open a thread on discourse to understand how predefined ids are working, in which situation they are useful and in which they are not. If the content of the thread will indicate that predefined ids are useful in our current situation. then I'll open up a second PR to rectify this.

What do you think ?

**Review by sterfield — APPROVED** [review]

**Comment by aanderse** [comment]
Sounds good to me.

**Comment by sterfield** [comment]
Here's the thread on discourse : https://discourse.nixos.org/t/predefined-uids-gids-when-to-use-them-and-when-not-to-use-them/956?u=sterfield

**Comment by sterfield** [comment]
I thought about it yesterday, and I found one use case where fixing the uid/gid could be beneficial.
Let's say you have a server running Nixos and a separate storage disk, where you store multiple files. Now, you are using an application that writes thousand of files under id, let's say 155. If you reinstall your server, it'll recreate the user for the application, but perhaps with a different uid. Now you have the same application, same user that cannot do anything on the storage disk, as all the files are owned by uid 155 that may not even exists on the server.

Predefined uid prevents this situation. Nothing that a good `find` can fix, but still.

You know what ? Merge this. I'll find the answers of my questions and will stop bothering you guys with that. After all, I'm the one having questions 😄

**Pull request merged**
